Race condition in locked tasks: wrong execution order

Re-activated by #210, this test produces surprising results:

https://travis-ci.org/github/peopledoc/procrastinate/jobs/687724543

It looks like jobs sometimes aren't ran in order, which contradicts one of our promises. We should really investigate this !

I've managed to reproduce the bug. Here is the pytest trace:

```
===================================================================================== test session starts =====================================================================================
platform linux -- Python 3.8.2, pytest-5.4.2, py-1.8.1, pluggy-0.13.1 -- /home/elemoine/.virtualenvs/procrastinate/bin/python3                                                                 
cachedir: .pytest_cache                                                                                                                                                                        
rootdir: /home/elemoine/src/procrastinate, inifile: setup.cfg                                                                                                                                  
plugins: mock-3.1.0, asyncio-0.12.0, cov-2.9.0, click-0.3                                                                                                                                      
collected 1 item                                                                                                                                                                               
                                                                                                                                                                                               
tests/acceptance/test_nominal.py::test_lock FAILED                                                                                                                                      [100%] 
                                                                                                                                                                                               
========================================================================================== FAILURES ===========================================================================================
__________________________________________________________________________________________ test_lock __________________________________________________________________________________________
                                                                                                                                                                                               
defer = <function defer.<locals>.func at 0x7f9fd08543a0>, running_worker = <function running_worker.<locals>.func at 0x7f9fcca56280>                                                           
                                                                                                                                                                                               
    def test_lock(defer, running_worker):                                                                                                                                                      
        """                                                                                                                                                                                    
        In this test, we launch 2 workers in two parallel threads, and ask them                                                                                                                
        both to process tasks with the same lock. We check that the second task is                                                                                                             
        not started before the first one was finished.                                                                                                                                         
        """                                                                                                                                                                                    
                                                                                                                                                                                               
        defer(                                                                                                                                                                                 
            "sleep_and_write",                                                                                                                                                                 
            lock="a",                                                                                                                                                                          
            sleep=0.5,                                                                                                                                                                         
            write_before="before-1",                                                                                                                                                           
            write_after="after-1",                                                                                                                                                             
        )                                                                                                                                                                                      
        defer(                                                                                                                                                                                 
            "sleep_and_write",                                                                                                                                                                 
            lock="a",                                                                                                                                                                          
            sleep=0.001,                                                                                                                                                                       
            write_before="before-2",                                                                                                                                                           
            write_after="after-2",                                                                                                                                                             
        )                                                                                                                                                                                      
        # Run the 2 workers concurrently                                                                                                                                                       
        process1 = running_worker(name="worker1")                                                                                                                                              
        process2 = running_worker(name="worker2")                                                                                                                                              
        time.sleep(2)                                                                                                                                                                          
        # And stop them
        process1.send_signal(signal.SIGINT)
        process2.send_signal(signal.SIGINT)
     
        # Gather their stdout
        stdout1, stderr1 = process1.communicate()
        print(stderr1)
        stdout2, stderr2 = process2.communicate()
        print(stderr2)
        stdout = stdout1 + stdout2
     
        # Sort the interesting lines by timestamp to reconstitute a consistent view
        lines = dict(
            line.split()[1:] for line in stdout.splitlines() if line.startswith("->")
        )
        lines = sorted(lines, key=lines.get)
     
        # Check that it all happened in order
>       assert lines == ["before-1", "after-1", "before-2", "after-2"]
E       AssertionError: assert ['before-2', 'after-2', 'before-1', 'after-1'] == ['before-1', 'after-1', 'before-2', 'after-2']
E         At index 0 diff: 'before-2' != 'before-1'
E         Full diff:
E         - ['before-1', 'after-1', 'before-2', 'after-2']
E         + ['before-2', 'after-2', 'before-1', 'after-1']

tests/acceptance/test_nominal.py:125: AssertionError
------------------------------------------------------------------------------------ Captured stdout call -------------------------------------------------------------------------------------
INFO:procrastinate.cli:Log level set to DEBUG
DEBUG:asyncio:Using selector: EpollSelector
DEBUG:procrastinate.app:Registering queue builtin
DEBUG:procrastinate.app:Registering queue default
DEBUG:procrastinate.app:Registering queue product_queue
DEBUG:procrastinate.app:Registering queue retry 
DEBUG:procrastinate.app:Registering queue lock_test
DEBUG:procrastinate.app:All tasks imported
INFO:procrastinate.worker.worker1:Starting worker on all queues
DEBUG:procrastinate.signals:Installing signal handler
DEBUG:procrastinate.worker.worker1:Waiting for new jobs on queues all queues
DEBUG:procrastinate.signals:Resetting previous signal handler
INFO:procrastinate.worker.worker1:Stop requested
INFO:procrastinate.worker.worker1:Stopped worker on all queues

INFO:procrastinate.cli:Log level set to DEBUG
DEBUG:asyncio:Using selector: EpollSelector
DEBUG:procrastinate.app:Registering queue builtin
DEBUG:procrastinate.app:Registering queue default
DEBUG:procrastinate.app:Registering queue product_queue
DEBUG:procrastinate.app:Registering queue retry 
DEBUG:procrastinate.app:Registering queue lock_test
DEBUG:procrastinate.app:All tasks imported
INFO:procrastinate.worker.worker2:Starting worker on all queues

------------------------------------------------------------------------------------ Captured stderr call -------------------------------------------------------------------------------------
INFO:procrastinate.cli:Log level set to DEBUG
DEBUG:asyncio:Using selector: EpollSelector
DEBUG:procrastinate.app:Registering queue builtin
DEBUG:procrastinate.app:Registering queue default
DEBUG:procrastinate.app:Registering queue product_queue
DEBUG:procrastinate.app:Registering queue retry 
DEBUG:procrastinate.app:Registering queue lock_test
DEBUG:procrastinate.app:All tasks imported
DEBUG:procrastinate.jobs:About to defer job tests.acceptance.app.sleep_and_write[None](sleep=0.5, write_before='before-1', write_after='after-1')
INFO:procrastinate.jobs:Deferred job tests.acceptance.app.sleep_and_write[None](sleep=0.5, write_before='before-1', write_after='after-1') with id: 1
INFO:procrastinate.cli:Log level set to DEBUG
DEBUG:asyncio:Using selector: EpollSelector
DEBUG:procrastinate.app:Registering queue builtin
DEBUG:procrastinate.app:Registering queue default
DEBUG:procrastinate.app:Registering queue product_queue
DEBUG:procrastinate.app:Registering queue retry 
DEBUG:procrastinate.app:Registering queue lock_test
DEBUG:procrastinate.app:All tasks imported
DEBUG:procrastinate.jobs:About to defer job tests.acceptance.app.sleep_and_write[None](sleep=0.001, write_before='before-2', write_after='after-2')
INFO:procrastinate.jobs:Deferred job tests.acceptance.app.sleep_and_write[None](sleep=0.001, write_before='before-2', write_after='after-2') with id: 2
====================================================================================== warnings summary =======================================================================================
tests/acceptance/test_nominal.py::test_lock
  /home/elemoine/.virtualenvs/procrastinate/lib/python3.8/site-packages/aiopg/pool.py:57: DeprecationWarning: The loop argument is deprecated since Python 3.8, and scheduled for removal in Py
thon 3.10.
    self._cond = asyncio.Condition(loop=self._loop)

tests/acceptance/test_nominal.py::test_lock
  /usr/lib64/python3.8/asyncio/locks.py:335: DeprecationWarning: The loop argument is deprecated since Python 3.8, and scheduled for removal in Python 3.10.
    lock = Lock(loop=loop)

tests/acceptance/test_nominal.py::test_lock
  /home/elemoine/.virtualenvs/procrastinate/lib/python3.8/site-packages/aiopg/connection.py:90: DeprecationWarning: The loop argument is deprecated since Python 3.8, and scheduled for removal
 in Python 3.10.
    self._notifies = asyncio.Queue(loop=self._loop)

tests/acceptance/test_nominal.py::test_lock
  /usr/lib64/python3.8/asyncio/queues.py:48: DeprecationWarning: The loop argument is deprecated since Python 3.8, and scheduled for removal in Python 3.10.
    self._finished = locks.Event(loop=loop)

tests/acceptance/test_nominal.py::test_lock
tests/acceptance/test_nominal.py::test_lock
  /home/elemoine/.virtualenvs/procrastinate/lib/python3.8/site-packages/aiopg/connection.py:207: DeprecationWarning: The loop argument is deprecated since Python 3.8, and scheduled for remova
l in Python 3.10.
    await asyncio.wait_for(self._waiter, timeout, loop=self._loop)

-- Docs: https://docs.pytest.org/en/latest/warnings.html

----------- coverage: platform linux, python 3.8.2-final-0 -----------
Name                               Stmts   Miss Branch BrPart  Cover   Missing
------------------------------------------------------------------------------
procrastinate/__init__.py             12      0      0      0   100%
procrastinate/__main__.py              5      5      2      0     0%   1-10
procrastinate/admin.py                17      7      8      0    48%   25, 57, 101, 142, 177-181
procrastinate/aiopg_connector.py     106     11     24      5    88%   31, 112->113, 113, 125->126, 126, 136->exit, 158->159, 159, 201-205, 248->249, 249, 253-255
procrastinate/app.py                  65      5      6      0    93%   199-201, 260, 264, 268
procrastinate/builtin_tasks.py         7      2      0      0    71%   28-29
procrastinate/cli.py                 156     39     34      7    67%   52-57, 86->91, 91, 145->exit, 146->149, 149, 194->195, 195, 227->228, 228, 233-234, 243->246, 246-249, 253->256, 256, 269-273, 300-308, 318-327, 337
procrastinate/connector.py            13      2      0      0    85%   10, 13
procrastinate/exceptions.py           15      1      0      0    93%   29
procrastinate/healthchecks.py         14      6      6      0    50%   10, 13-16, 19-23
procrastinate/job_context.py          36     11     10      1    65%   59->60, 60, 65-76
procrastinate/jobs.py                 57      2     10      2    94%   25->26, 26, 86->87, 87
procrastinate/metadata.py              5      0      0      0   100%
procrastinate/retry.py                37     14     14      1    59%   23-28, 82-92, 102->103, 103
procrastinate/schema.py               18      1      2      0    95%   23
procrastinate/shell.py                36     21     12      0    31%   7-8, 12-19, 27-28, 32, 47-50, 62-64, 81-83, 99, 110
procrastinate/signals.py              48     11     12      5    70%   27->28, 28-33, 40->41, 41-44, 48-49, 63->64, 64, 75->79, 79-80, 92->95
procrastinate/sql/__init__.py         18      0      0      0   100%
procrastinate/store.py                32      7     10      1    67%   9->12, 12, 50-56, 65-70
procrastinate/tasks.py                46     12      8      2    70%   16-21, 34->35, 35, 37->38, 38, 105-107, 167-170
procrastinate/testing.py             102     69     48      0    22%   31-33, 39-44, 52-54, 57, 60, 63, 68, 73-74, 79-100, 104, 108, 117-132, 137-148, 151, 162-172, 175, 178, 181-186, 189-194, 197-202, 205-206
procrastinate/types.py                 3      0      0      0   100%
procrastinate/utils.py                70     19     22      6    66%   19->20, 20, 24-25, 29-30, 32->33, 33, 45->46, 46-50, 94->95, 95, 109->111, 111-116, 130->131, 131-132, 141-143
procrastinate/worker.py              150     39     34      9    71%   39->42, 42, 66->70, 115->116, 116, 131-132, 149-155, 172->173, 173, 178-196, 213->214, 214, 217->218, 218, 219->220, 220-238, 265->270, 279->280, 280
------------------------------------------------------------------------------
TOTAL                               1068    284    262     39    67%
Coverage HTML written to dir htmlcov

=================================================================================== short test summary info ===================================================================================
FAILED tests/acceptance/test_nominal.py::test_lock - AssertionError: assert ['before-2', 'after-2', 'before-1', 'after-1'] == ['before-1', 'after-1', 'before-2', 'after-2']
================================================================================ 1 failed, 6 warnings in 4.08s ================================================================================
```

--- a/docs/discussions.rst
+++ b/docs/discussions.rst
@@ -80,10 +80,10 @@ About locks
 -----------
 
 Let's say we have a :term:`task` that writes a character at the end of a file after
-waiting for a random amount of time. This represents a real world problem where tasks
+waiting for a random amount of time. This represents a real world problem where jobs
 take an unforeseeable amount of time and share resources like a database.
 
-We launch 4 tasks respectively writing ``a``, ``b``, ``c`` and ``d``. We would expect
+We launch 4 jobs respectively writing ``a``, ``b``, ``c`` and ``d``. We would expect
 the file to contain ``abcd``, but it's not the case, for example maybe it's ``badc``.
 The jobs were taken from the queue in order, but because we have several workers, the
 jobs were launched in parallel and because their duration is random, the final result
@@ -91,33 +91,37 @@ pretty much is too.
 
 We can solve this problem by using locks. Procrastinate gives us two guarantees:
 
-- Tasks are consumed in creation order. When a worker requests a task, it will always
-  receive the oldest available task. Unavailable tasks, either locked, scheduled for the
-  future or in a queue that the worker doesn't listen to, will be ignored.
-- If a group of tasks share the same lock, then only one can be executed at a time.
+- Jobs are consumed in creation order. When a worker requests a job, it can recieve
+  a job with a lock, or a job without a lock. If there is a lock, then the received
+  job will be the oldest one with that lock. If the oldest job awaiting execution is
+  not available for this worker (either it's on a queue that this worker doesn't
+  listen to, or it's scheduled in the future), then jobs with this lock will not be
+  considered.
+- If a group of jobs share the same lock, then only one can be executed at a time.
 
-These two facts allow us to draw the following conclusion for our 4 letter tasks from
-above. If our 4 tasks share the same lock (for example, the name of the file we're
+These two facts allow us to draw the following conclusion for our 4 letter jobs from
+above. If our 4 jobs share the same lock (for example, the name of the file we're
 writing to):
 
-- The 4 tasks will be started in order;
-- A task will not start before the previous one is finished.
+- The 4 jobs will be started in order;
+- A job will not start before the previous one is finished.
 
 This says we can safely expect the file to contain ``abcd``.
 
 Note that Procrastinate will use PostgreSQL to search the jobs table for suitable jobs.
-Even if the database contains a high proportion of locked tasks, this will barely affect
-Procrastinates's capacity to quickly find the free tasks.
+Even if the database contains a high proportion of locked jobs, this will barely affect
+Procrastinates's capacity to quickly find the free jobs.
 
 A good string identifier for the lock is a string identifier of the shared resource,
 UUIDs are well suited for this. If multiple resources are implicated, a combination of
 their identifiers could be used (there's no hard limit on the length of a lock string,
 but stay reasonable).
 
-A task can only take a single lock so there's no dead-lock scenario possible where two
-running tasks are waiting one another. That being said, if a worker dies with a lock, it
-will be up to you to free it. If the task fails but the worker survives though, the
-lock will be freed.
+A job can only take a single lock so there's no dead-lock scenario possible where two
+running jobs are waiting for one another. If a worker is killed without ending its job,
+following jobs with the same lock will not run until the interrupted job is either
+manually set to "failed" or "succeeded". If a job simply fails, following jobs with the
+same locks may run.
 
 For a more practical approach, see `howto/locks`.
 
@@ -228,21 +232,6 @@ For critical requests, we tend to using PostgreSQL procedures where we could do 
 thing directly with queries. This is so that the database is solely responsible for
 consistency, and would allow us to have the same behavior if someone were to write
 a procrastinate compatible client, in Python or in another language altogether.
-
-The ``procrastinate_job_locks`` table
-^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
-
-We could have used PostgreSQL's `advisory locks`_, and we choose to kinda "reimplement
-the wheel" with a dedicated table. This is because we made the choice that if a worker
-dies holding a lock, we'd rather have a human examine the situation and manually free
-the lock than having the lock been automatically freed, and fail our locks consistency
-guarantee.
-
-.. _`advisory locks`: https://www.postgresql.org/docs/10/explicit-locking.html#ADVISORY-LOCKS
-
-So far, Procrastinate implements async job :term:`deferring <defer>`, and async job
-executing but not in parallel, meaning it can run jobs written as a coroutine, but it
-will only execute one job at a time.
 
 Why is Procrastinate asynchronous at core?
 ------------------------------------------

--- a/procrastinate/sql/migrations/delta_0.9.0_001_close_fetch_job_race_condition.sql
+++ b/procrastinate/sql/migrations/delta_0.9.0_001_close_fetch_job_race_condition.sql
@@ -11,12 +11,14 @@ BEGIN
     WITH candidate AS (
         SELECT jobs.*
             FROM procrastinate_jobs AS jobs
-            LEFT JOIN procrastinate_jobs AS earlier_jobs
-            ON (earlier_jobs.lock = jobs.lock
-                AND earlier_jobs.status IN ('todo', 'doing')
-                AND earlier_jobs.id < jobs.id)
             WHERE
-                earlier_jobs.id IS NULL  -- reject jobs that have earlier jobs with the same lock
+                -- reject the job if its lock has earlier jobs
+                NOT EXISTS (
+                    SELECT 1
+                        FROM procrastinate_jobs AS earlier_jobs
+                        WHERE earlier_jobs.lock = jobs.lock
+                            AND earlier_jobs.status IN ('todo', 'doing')
+                            AND earlier_jobs.id < jobs.id)
                 AND jobs.status = 'todo'
                 AND (target_queue_names IS NULL OR jobs.queue_name = ANY( target_queue_names ))
                 AND (jobs.scheduled_at IS NULL OR jobs.scheduled_at <= now())

--- a/procrastinate/sql/migrations/delta_0.9.0_001_close_fetch_job_race_condition.sql
+++ b/procrastinate/sql/migrations/delta_0.9.0_001_close_fetch_job_race_condition.sql
@@ -1,0 +1,46 @@
+DROP TABLE IF EXISTS procrastinate_job_locks;
+
+CREATE UNIQUE INDEX procrastinate_jobs_lock_idx ON procrastinate_jobs (lock) WHERE status = 'doing';
+
+CREATE OR REPLACE FUNCTION procrastinate_fetch_job(target_queue_names character varying[]) RETURNS procrastinate_jobs
+    LANGUAGE plpgsql
+    AS $$
+DECLARE
+	found_jobs procrastinate_jobs;
+BEGIN
+    WITH candidate AS (
+        SELECT jobs.*
+            FROM procrastinate_jobs AS jobs
+            LEFT JOIN procrastinate_jobs AS earlier_jobs
+            ON (earlier_jobs.lock = jobs.lock
+                AND earlier_jobs.status IN ('todo', 'doing')
+                AND earlier_jobs.id < jobs.id)
+            WHERE
+                earlier_jobs.id IS NULL  -- reject jobs that have earlier jobs with the same lock
+                AND jobs.status = 'todo'
+                AND (target_queue_names IS NULL OR jobs.queue_name = ANY( target_queue_names ))
+                AND (jobs.scheduled_at IS NULL OR jobs.scheduled_at <= now())
+            ORDER BY jobs.id ASC LIMIT 1
+            FOR UPDATE OF jobs SKIP LOCKED
+    )
+    UPDATE procrastinate_jobs
+        SET status = 'doing'
+        FROM candidate
+        WHERE procrastinate_jobs.id = candidate.id
+        RETURNING procrastinate_jobs.* INTO found_jobs;
+
+	RETURN found_jobs;
+END;
+$$;
+
+CREATE OR REPLACE FUNCTION procrastinate_finish_job(job_id integer, end_status procrastinate_job_status, next_scheduled_at timestamp with time zone) RETURNS void
+    LANGUAGE plpgsql
+    AS $$
+BEGIN
+    UPDATE procrastinate_jobs
+    SET status = end_status,
+        attempts = attempts + 1,
+        scheduled_at = COALESCE(next_scheduled_at, scheduled_at)
+    WHERE id = job_id;
+END;
+$$;

--- a/procrastinate/sql/schema.sql
+++ b/procrastinate/sql/schema.sql
@@ -56,12 +56,14 @@ BEGIN
     WITH candidate AS (
         SELECT jobs.*
             FROM procrastinate_jobs AS jobs
-            LEFT JOIN procrastinate_jobs AS earlier_jobs
-            ON (earlier_jobs.lock = jobs.lock
-                AND earlier_jobs.status IN ('todo', 'doing')
-                AND earlier_jobs.id < jobs.id)
             WHERE
-                earlier_jobs.id IS NULL  -- reject jobs that have earlier jobs with the same lock
+                -- reject the job if its lock has earlier jobs
+                NOT EXISTS (
+                    SELECT 1
+                        FROM procrastinate_jobs AS earlier_jobs
+                        WHERE earlier_jobs.lock = jobs.lock
+                            AND earlier_jobs.status IN ('todo', 'doing')
+                            AND earlier_jobs.id < jobs.id)
                 AND jobs.status = 'todo'
                 AND (target_queue_names IS NULL OR jobs.queue_name = ANY( target_queue_names ))
                 AND (jobs.scheduled_at IS NULL OR jobs.scheduled_at <= now())

--- a/procrastinate/sql/schema.sql
+++ b/procrastinate/sql/schema.sql
@@ -35,15 +35,16 @@ CREATE TABLE procrastinate_jobs (
 -- this prevents from having several jobs with the same queueing lock in the "todo" state
 CREATE UNIQUE INDEX procrastinate_jobs_queueing_lock_idx ON procrastinate_jobs (queueing_lock) WHERE status = 'todo';
 
+-- this prevents from having several jobs with the same lock in the "doing" state
+CREATE UNIQUE INDEX procrastinate_jobs_lock_idx ON procrastinate_jobs (lock) WHERE status = 'doing';
+
+CREATE INDEX ON procrastinate_jobs(queue_name);
+
 CREATE TABLE procrastinate_events (
     id BIGSERIAL PRIMARY KEY,
     job_id integer NOT NULL REFERENCES procrastinate_jobs ON DELETE CASCADE,
     type procrastinate_job_event_type,
     at timestamp with time zone DEFAULT NOW() NULL
-);
-
-CREATE TABLE procrastinate_job_locks (
-    object text PRIMARY KEY
 );
 
 CREATE FUNCTION procrastinate_fetch_job(target_queue_names character varying[]) RETURNS procrastinate_jobs
@@ -52,28 +53,26 @@ CREATE FUNCTION procrastinate_fetch_job(target_queue_names character varying[]) 
 DECLARE
 	found_jobs procrastinate_jobs;
 BEGIN
-	WITH potential_job AS (
-		SELECT procrastinate_jobs.*
-			FROM procrastinate_jobs
-			LEFT JOIN procrastinate_job_locks ON procrastinate_job_locks.object = procrastinate_jobs.lock
-			WHERE (target_queue_names IS NULL OR queue_name = ANY( target_queue_names ))
-			  AND procrastinate_job_locks.object IS NULL
-			  AND status = 'todo'
-			  AND (scheduled_at IS NULL OR scheduled_at <= now())
-            ORDER BY id ASC
-			FOR UPDATE OF procrastinate_jobs SKIP LOCKED LIMIT 1
-	), lock_object AS (
-		INSERT INTO procrastinate_job_locks
-			SELECT lock FROM potential_job
-            ON CONFLICT DO NOTHING
-            RETURNING object
-	)
-	UPDATE procrastinate_jobs
-		SET status = 'doing'
-		FROM potential_job, lock_object
-        WHERE lock_object.object IS NOT NULL
-		AND procrastinate_jobs.id = potential_job.id
-		RETURNING procrastinate_jobs.* INTO found_jobs;
+    WITH candidate AS (
+        SELECT jobs.*
+            FROM procrastinate_jobs AS jobs
+            LEFT JOIN procrastinate_jobs AS earlier_jobs
+            ON (earlier_jobs.lock = jobs.lock
+                AND earlier_jobs.status IN ('todo', 'doing')
+                AND earlier_jobs.id < jobs.id)
+            WHERE
+                earlier_jobs.id IS NULL  -- reject jobs that have earlier jobs with the same lock
+                AND jobs.status = 'todo'
+                AND (target_queue_names IS NULL OR jobs.queue_name = ANY( target_queue_names ))
+                AND (jobs.scheduled_at IS NULL OR jobs.scheduled_at <= now())
+            ORDER BY jobs.id ASC LIMIT 1
+            FOR UPDATE OF jobs SKIP LOCKED
+    )
+    UPDATE procrastinate_jobs
+        SET status = 'doing'
+        FROM candidate
+        WHERE procrastinate_jobs.id = candidate.id
+        RETURNING procrastinate_jobs.* INTO found_jobs;
 
 	RETURN found_jobs;
 END;
@@ -83,14 +82,11 @@ CREATE FUNCTION procrastinate_finish_job(job_id integer, end_status procrastinat
     LANGUAGE plpgsql
     AS $$
 BEGIN
-	WITH finished_job AS (
-		UPDATE procrastinate_jobs
-        SET status = end_status,
-            attempts = attempts + 1,
-            scheduled_at = COALESCE(next_scheduled_at, scheduled_at)
-        WHERE id = job_id RETURNING lock
-	)
-	DELETE FROM procrastinate_job_locks WHERE object = (SELECT lock FROM finished_job);
+    UPDATE procrastinate_jobs
+    SET status = end_status,
+        attempts = attempts + 1,
+        scheduled_at = COALESCE(next_scheduled_at, scheduled_at)
+    WHERE id = job_id;
 END;
 $$;
 
@@ -159,8 +155,6 @@ BEGIN
 	RETURN NEW;
 END;
 $$;
-
-CREATE INDEX ON procrastinate_jobs(queue_name);
 
 CREATE TRIGGER procrastinate_jobs_notify_queue
     AFTER INSERT ON procrastinate_jobs

